### PR TITLE
Add California Law HLD

### PR DESCRIPTION
**Pull request by andriydnvd**

### CODE REVIEW

File: doc/california_law/California_law.md [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+See linux [3rd Party Components](#rdPartyComponents) for more description.
+
+
+##### Flow diagram:
+
+![California law flow](California_law_diagram.jpg)
+
+####  1.7.1. <a name='Flowdescription'></a>Flow description
+Feature is expected to be disabled by default.
+Build flag CHANGE_DEFAULT_PASWD will be used to set feature state.
```
Comment (jeffhtgt) [inline doc/california_law/California_law.md:100]: Could we change the compile flag to "CHANGE_DEFAULT_PASSWORD" for consistency?

File: doc/california_law/California_law.md [hunk 2]
```diff
@@ -0,0 +1,236 @@
+# California Law Design #
+
+##  1. <a name='TableofContent'></a>Table of Content
```
Comment (lihuay) [inline doc/california_law/California_law.md:3]: can we use a more descriptive name? "Default Credential Management for California Law XXX Conformance"?

File: doc/california_law/California_law.md [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+###  1.11. <a name='WarmbootandFastbootDesignImpact'></a>Warmboot and Fastboot Design Impact
+The feature can be triggered after sonic upgrade and warm reboot and feature doesn't affect trafic.
+
+###  1.12. <a name='RestrictionsLimitations'></a>Restrictions/Limitations
+The California law feature is not supported on remote AAA.
+
+LDAP/Radius/Tacacs is under customer responsibility.
+
+###  1.13. <a name='UpgradeFlow'></a>Upgrade Flow
+After upgrade to new image with soni-istall 1st boot flow is triggered and users will be forced to change their passwords
```
Comment (lihuay) [inline doc/california_law/California_law.md:152]: install

**Review by andriydnvd — COMMENTED** [review]
Added changes accordingly to review comments

**Review by andriydnvd — COMMENTED** [review]
Fixes done

File: doc/california_law/California_law.md [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+###  1.11. <a name='WarmbootandFastbootDesignImpact'></a>Warmboot and Fastboot Design Impact
+The feature can be triggered after sonic upgrade and warm reboot and feature doesn't affect trafic.
+
+###  1.12. <a name='RestrictionsLimitations'></a>Restrictions/Limitations
+The California law feature is not supported on remote AAA.
+
+LDAP/Radius/Tacacs is under customer responsibility.
+
+###  1.13. <a name='UpgradeFlow'></a>Upgrade Flow
+After install new image with sonic-istall 1st boot flow is triggered and users will be forced to change their passwords
```
Comment (liuh-80) [inline doc/california_law/California_law.md:152]: Typo? sonic-istall

File: doc/california_law/California_law.md [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+###  1.4. <a name='Overview'></a>Overview
+
+Most users won't change the password because of various reasons: forgot to change, don't realize default password should be changed and other. As a result
+default passwords in connected devices is one of a primary attack method used by malicious actors. The main motivation and functionality of the feature is to
+force user to change default password on first login.
+
+###  1.5. <a name='Requirements'></a>Requirements
+
+ - Force change of default password on first login
+ - Support several default users
```
Comment (liuh-80) [inline doc/california_law/California_law.md:67]: Suggest clearify 'default user', as my understand we don't need expire users who's shell is /bin/false or /usr/sbin/nologin.

**Review by andriydnvd — COMMENTED** [review]
Updated doc, pls check

**Review by lihuay — APPROVED** [review]

**Comment by liat-grozovik** [comment]
@lihuay would you mind to merge or should we wait for additional reviewers?

**Comment by liuh-80** [comment]
@andriydnvd, what's the plan implement feature in sonic-buildimage repo?
Currently there are sonic-mgmt test case, and some validation pipelines depends on default password. so you may also need improve those test case and validation pipelines to handle password change.

**Comment by zhangyanzhao** [comment]
@andriydnvd can you please add the code PR into this HLD PR by referring to #806 ? Thanks.

**Comment by andriydnvd** [comment]
This PR contains California-SB237 HLD

| PR title | state | context |
| ---------------------------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------------------------------------------ |
| [[sonic-buildimage] Added support for California-SB237](https://github.com/Azure/sonic-buildimage/pull/12678) | GitHub issue/pull request detail | GitHub pull request check contexts |

File: doc/California-SB237/California-SB237.md [hunk 6]
```diff
@@ -0,0 +1,238 @@
...
+	You must change your password now and login again!
+	Changing password for admin.
+	Current password:
+	New password:
+	Retype new password:
+	The password has not been changed.
+	New password:
+	Retype new password:
+	passwd: password updated successfully
+	Connection to sonic-switch closed.
```
Comment (qiluo-msft) [inline doc/California-SB237/California-SB237.md:211]: >Connection to sonic-switch closed. [](http://example.com/codeflow?start=1&length=34)

Is it true that `updated successfully` will logout the user immediately?

File: doc/California-SB237/California-SB237.md [hunk 7]
```diff
@@ -0,0 +1,238 @@
...
+default passwords in connected devices is one of a primary attack method used by malicious actors. The main motivation and functionality of the feature is to
+force user to change default password on first login.
+
+###  1.5. <a name='Requirements'></a>Requirements
+
+ - Force change of default password on first login
+ - Support several default users
+ - Force to change password after image update
+ - Don't affect [Password hardening feature](https://github.com/sonic-net/SONiC/blob/master/doc/passw_hardening/hld_password_hardening.md)
+ - Expire password only for user that can do login. (login shell is /bin/bash or /bin/sh)
```
Comment (qiluo-msft) [inline doc/California-SB237/California-SB237.md:71]: >bash [](http://example.com/codeflow?start=73&length=4)

how about zsh or fish?

File: doc/California-SB237/California-SB237.md [hunk 8]
```diff
@@ -0,0 +1,238 @@
...
+Common approach to force user to change a password is making user password expired.
+
+The feature will use native Linux tool and linux lib: chage and pam_unix_account.so
+
+chage: support requirement of PW aging, change user password expire information
+
+pam_unix_account module retrieves password aging information and verifies that the password and the user's account have not expired.
+
+Note:
+See linux [3rd Party Components](#rdPartyComponents) for more description.
```
Comment (qiluo-msft) [inline doc/California-SB237/California-SB237.md:93]: >(#rdPartyComponents) [](http://example.com/codeflow?start=32&length=20)

This markdown link does not work.

File: doc/California-SB237/California-SB237.md [hunk 9]
```diff
@@ -0,0 +1,238 @@
...
+
+
+##### Flow diagram:
+
+![California law flow](California-SB237_diagram.jpg)
```
Comment (qiluo-msft) [inline doc/California-SB237/California-SB237.md:98]: In diagram:
Default user_1 -> user_1

File: doc/California-SB237/California-SB237.md [hunk 10]
```diff
@@ -0,0 +1,238 @@
...
+	passwd: password updated successfully
+	Connection to sonic-switch closed.
+
+
+###  1.15. <a name='rdPartyComponents'></a>3rd Party Components
+In this section, you can find options used for password expiring with chage and pwhistory
+
+####  1.15.1. <a name='PWForceExpire'></a>PW Force Expiration
+
+Tool for password expiration:
```
Comment (qiluo-msft) [inline doc/California-SB237/California-SB237.md:219]: >expiration [](http://example.com/codeflow?start=18&length=10)

I am considering more use cases which requiring CLI:
1. Force all user password expiring
2. Force all user password not expiring any more

In real life, the manufactory and customers are login the device for different purpose:
1. the image is build with CHANGE_DEFAULT_PASSWORD turn on.
2. the manufactory test team will login and change password, and continue test
3. (missing) At the end of test, the manufactory test team will force all user password expiring
4. customers got the device, and login, and change password
5. customer start use it in normal way.

**Pull request merged**